onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -0,0 +1,706 @@
+    const std::string& post_transform,
+    std::vector<int64_t> target_class_ids,
+    std::vector<int64_t> target_class_nodeids,
+    std::vector<int64_t> target_class_treeids,
+    std::vector<OTYPE> target_class_weights) {
+  init_c(aggregate_function, base_values, n_targets_or_classes,
+         nodes_falsenodeids, nodes_featureids, nodes_hitrates,
+         nodes_missing_value_tracks_true, nodes_modes,
+         nodes_nodeids, nodes_treeids, nodes_truenodeids,
+         nodes_values, post_transform, target_class_ids,
+         target_class_nodeids, target_class_treeids,
+         target_class_weights);
+}
+
+template <typename ITYPE, typename OTYPE>
+void TreeEnsembleCommon<ITYPE, OTYPE>::init_c(
+    const std::string& aggregate_function,
+    const std::vector<OTYPE>& base_values,
+    int64_t n_targets_or_classes,
+    const std::vector<int64_t>& nodes_falsenodeids,
+    const std::vector<int64_t>& nodes_featureids,
+    const std::vector<OTYPE>& nodes_hitrates,
+    const std::vector<int64_t>& nodes_missing_value_tracks_true,
+    const std::vector<std::string>& nodes_modes,
+    const std::vector<int64_t>& nodes_nodeids,
+    const std::vector<int64_t>& nodes_treeids,
+    const std::vector<int64_t>& nodes_truenodeids,
+    const std::vector<OTYPE>& nodes_values,
+    const std::string& post_transform,
+    const std::vector<int64_t>& target_class_ids,
+    const std::vector<int64_t>& target_class_nodeids,
+    const std::vector<int64_t>& target_class_treeids,
+    const std::vector<OTYPE>& target_class_weights) {
+  aggregate_function_ = ::onnxruntime::ml::MakeAggregateFunction(aggregate_function);
+  post_transform_ = ::onnxruntime::ml::MakeTransform(post_transform);
+  base_values_ = base_values;
+  n_targets_or_classes_ = n_targets_or_classes;
+  max_tree_depth_ = 1000;
+
+  // additional members
+  std::vector<NODE_MODE> cmodes(nodes_modes.size());
+  same_mode_ = true;
+  int fpos = -1;
+  for (size_t i = 0; i < nodes_modes.size(); ++i) {
+    cmodes[i] = ::onnxruntime::ml::MakeTreeNodeMode(nodes_modes[i]);
+    if (cmodes[i] == NODE_MODE::LEAF)
+      continue;
+    if (fpos == -1) {
+      fpos = static_cast<int>(i);
+      continue;
+    }
+    if (cmodes[i] != cmodes[fpos])
+      same_mode_ = false;
+  }
+
+  // filling nodes
+
+  n_nodes_ = nodes_treeids.size();
+  nodes_ = new TreeNodeElement<OTYPE>[static_cast<int>(n_nodes_)];
+  roots_.clear();
+  std::map<TreeNodeElementId, TreeNodeElement<OTYPE>*> idi;
+  size_t i;
+
+  for (i = 0; i < nodes_treeids.size(); ++i) {
+    TreeNodeElement<OTYPE>* node = nodes_ + i;
+    node->id.tree_id = static_cast<int>(nodes_treeids[i]);
+    node->id.node_id = static_cast<int>(nodes_nodeids[i]);
+    node->feature_id = static_cast<int>(nodes_featureids[i]);
+    node->value = nodes_values[i];
+    node->hitrates = i < nodes_hitrates.size() ? nodes_hitrates[i] : -1;
+    node->mode = cmodes[i];
+    node->is_not_leave = node->mode != NODE_MODE::LEAF;
```

Comment:
Would be good if something validated all these things are the correct size, or at() was used to do that.